Assets/Scripts/MovePlayer.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class MovePlayer : MonoBehaviour
+{
+    #region SerializeFields
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a slow movement (through sand, etc.).")]
+    private float slowMoveTime = .75f;
+    [SerializeField]
+    [Tooltip("Duration, in seconds, of a single movement.")]
+    private float moveTime = .5f;
+    [SerializeField]
+    [Tooltip("Threshold to stop movement coroutine and snap to position.")]
+    private float movementMargin = .05f;
+    [SerializeField]
+    [Tooltip("Point from which to raycast to check collision before moving.")]
+    private Transform collisionRaycastPoint;
+    [SerializeField]
+    [Tooltip("Layer(s) the player collides with when moving.")]
+    private LayerMask collisionLayerMask;
+    [SerializeField]
+    [Tooltip("Layer that slows the player down when moving.")]
+    private LayerMask slowTerrainLayerMask;
+    [SerializeField]
+    [Tooltip("Tag of terrain that slows the player down when moving.")]
+    private string slowTerrainTag;
+    #endregion
+
+    /// <summary>
+    /// Defines directions in an enum for legible array reference
+    /// </summary>
+    private enum directions { Down, Up, Left, Right};
+
+    /// <summary>
+    /// The animator attached to the player
+    /// </summary>
+    private Animator playerAnimator;
+    /// <summary>
+    /// Is the player allowed to move? Will be false if paused, in a cutscene, etc.
+    /// </summary>
+    private bool canMove = true;
+    /// <summary>
+    /// Has the player received input to move?
+    /// </summary>
+    private bool shouldMove;
+    /// <summary>
+    /// Is the player in slow terrain?
+    /// </summary>
+    private bool slowMove;
+    /// <summary>
+    /// Is the movement coroutine currently running?
+    /// </summary>
+    private bool moving;
+    /// <summary>
+    /// Which direction is the player trying to move?
+    /// </summary>
+    private directions currentDirection;
+    /// <summary>
+    /// An array for quick reference of directions, in conjunction with the enum
+    /// This prevents the need to write four separate coroutines
+    /// </summary>
+    private Vector3[] movementVectors;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        playerAnimator = GetComponent<Animator>();
+        movementVectors = new Vector3[4] { Vector2.down, Vector2.up, Vector2.left, Vector2.right };
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        GetInput();
+    }
+
+    private void FixedUpdate()
+    {
+        if (canMove && shouldMove && !moving)
+            StartCoroutine(MovementRoutine());
+    }
+
+    private void OnTriggerEnter2D(Collider2D collision)
+    {
+        if (collision.gameObject.tag == slowTerrainTag)
+            slowMove = !slowMove;
```

Comment:
won't this toggle my slow terrain status each time I enter a slow terrain? If the player is moving across consecutive slow terrain tiles, will this make them alternate between moving slow and moving normal when they should be moving slow the whole time?

Reply:
What's the intent of this code, I guess is a good way to start. Put it in a method with a summary that clarifies the intent.

Reply:
I can put it in a method. The annoying thing is that it actually doesn't toggle when you walk across slow terrain. Because of the composite collider making a large bounding box around the entire section of terrain, the raycasting method only worked when entering or exiting the terrain. In my testing, this ensured that they continued to move slowly while in slow terrain, then returned to normal speed when they left. The raycast makes them move slowly on their first step into the slow terrain, before they've hit the collider.